tensorflow/compiler/xla/service/cpu/onednn_rewriter.cc

```diff
@@ -0,0 +1,125 @@
+#if defined(INTEL_MKL) && defined(ENABLE_ONEDNN_V3)
+
+#include "tensorflow/compiler/xla/service/cpu/onednn_rewriter.h"
+
+#include "tensorflow/compiler/xla/hlo/ir/dfs_hlo_visitor_with_default.h"
+#include "tensorflow/compiler/xla/hlo/ir/hlo_instruction.h"
+#include "tensorflow/compiler/xla/service/cpu/backend_config.pb.h"
+#include "tensorflow/compiler/xla/service/cpu/onednn_memory_util.h"
+#include "tensorflow/compiler/xla/service/pattern_matcher.h"
+#include "tensorflow/compiler/xla/status_macros.h"
+
+namespace xla {
+namespace cpu {
+
+namespace {
+namespace m = match;
+
+Status ValidateDotDimensionNumbers(const DotDimensionNumbers& dim_numbers) {
+  // Checks some invariants that do not hold in general, but DotDecomposer
+  // should have established for us.
+  TF_RET_CHECK(dim_numbers.lhs_contracting_dimensions_size() == 1);
+  std::vector<int64_t> batch_dim_numbers(
+      dim_numbers.lhs_batch_dimensions_size());
+  absl::c_iota(batch_dim_numbers, 0);
+  TF_RET_CHECK(
+      absl::c_equal(batch_dim_numbers, dim_numbers.lhs_batch_dimensions()));
+  TF_RET_CHECK(
+      absl::c_equal(batch_dim_numbers, dim_numbers.rhs_batch_dimensions()));
+  return OkStatus();
+}
+
+}  // namespace
+
+class OneDnnRewriterVisitor : public DfsHloRewriteVisitor {
+ public:
+  // Matches patterns for possible MatMul fusions that are supported by oneDNN
+  // library. Matched hlo instruction(s) are replaced by custom call.
+  Status HandleDot(HloInstruction* instr) override {
+    // Currently, blocking control dependencies
+    if (instr->HasControlDependencies()) return OkStatus();
+    HloInstruction* dot_instr;
+    auto pattern = m::Op(&dot_instr).WithOpcode(HloOpcode::kDot);
+    if (!Match(instr, pattern)) return OkStatus();
+
+    // The rewrite pass runs after dot-decomposition pass. Adjust
+    // the rewrite condition when the rewrite pass is moved before
+    // dot-decomposition pass.
+
+    // Currently, we rewrite when the data type is F32 or BF16. Note we do not
+    // need to check equality of contraction dim-size of the operands. HLO
+    // verifier already does the job. We however, need to check if contraction
+    // is over only 1 dimension (a.k.a. K dimension in matrix-multiplication
+    // parlance). We also restrict batch dimensions of the operands mathces.
+    if (auto dtype = dot_instr->shape().element_type();
+        !(dtype == F32 || dtype == BF16))
+      return OkStatus();
+    auto dot_dim_numbers = dot_instr->dot_dimension_numbers();
+    TF_RETURN_IF_ERROR(ValidateDotDimensionNumbers(dot_dim_numbers));
+    const Shape& lhs_shape = dot_instr->operand(0)->shape();
+    const Shape& rhs_shape = dot_instr->operand(1)->shape();
+    const Shape& output_shape = dot_instr->shape();
+    bool should_rewrite = true;
+    should_rewrite &= (lhs_shape.rank() == rhs_shape.rank());
+    should_rewrite &= (rhs_shape.rank() == output_shape.rank());
+    // OneDNN only  supports rank >=2 and <= kOneDnnMaxNDims.
```

Comment:
Nit: Making it easier to read.
```suggestion
    // OneDNN only  supports 2 <= rank <= kOneDnnMaxNDims.
```